experimental/iterators/lib/Conversion/IteratorsToStandard/IteratorsToStandard.cpp

```diff
@@ -0,0 +1,159 @@
+public:
+  IteratorsTypeConverter() {
+    addConversion([](Type type) { return type; });
+    addConversion(convertIteratorType);
+  }
+
+  static Optional<Type> convertIteratorType(Type type) {
+    if (type.isa<iterators::IteratorType>())
+      return LLVM::LLVMPointerType::get(IntegerType::get(type.getContext(), 8));
+    return llvm::None;
+  }
+};
+
+/// Returns or creates a function declaration at the module of the provided
+/// original op,
+FuncOp lookupOrCreateFuncOp(llvm::StringRef fnName, FunctionType fnType,
+                            Operation *op, PatternRewriter &rewriter) {
+  ModuleOp module = op->getParentOfType<ModuleOp>();
+
+  // Return function if already declared
+  if (FuncOp funcOp = module.lookupSymbol<mlir::FuncOp>(fnName))
+    return funcOp;
+
+  // Add new declaration at the end of the module
+  OpBuilder::InsertionGuard guard(rewriter);
+  rewriter.setInsertionPoint(module.getBody(),
+                             std::prev(module.getBody()->end()));
+  FuncOp funcOp = rewriter.create<FuncOp>(op->getLoc(), fnName, fnType);
+  funcOp.setPrivate();
+  return funcOp;
+}
+
+/// Replaces a instances of a certain IteratorOp with a call to the given
+/// external constructor as well as a call to the given destructor at the end of
+/// the block.
+struct IteratorConversionPattern : public ConversionPattern {
+  IteratorConversionPattern(TypeConverter &typeConverter, MLIRContext *context,
+                            StringRef rootName, StringRef constructorName,
+                            StringRef destructorName,
+                            PatternBenefit benefit = 1)
+      : ConversionPattern(typeConverter, rootName, benefit, context),
+        constructorName(constructorName), destructorName(destructorName) {}
+
+  LogicalResult
+  matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const override {
+    // Convert result types
+    llvm::SmallVector<Type, 4> resultTypes;
+    if (typeConverter->convertTypes(op->getResultTypes(), resultTypes).failed())
+      return failure();
+    assert(resultTypes.size() <= 1 &&
+           "Iterators may have only one output (and sinks have none).");
+
+    // Constructor (aka "iteratorsMake*Operator")
+
+    // Look up or declare function symbol
+    auto const fnType =
+        FunctionType::get(getContext(), TypeRange(operands), resultTypes);
+    FuncOp funcOp = lookupOrCreateFuncOp(constructorName, fnType, op, rewriter);
+
+    // Replace op with call to function
+    func::CallOp callOp =
+        rewriter.replaceOpWithNewOp<func::CallOp>(op, funcOp, operands);
+
+    // Destructor (aka "iteratorsDestroy*Operator")
+
+    // No destructor necessary for sinks
+    if (resultTypes.empty())
+      return success();
+    assert(resultTypes.size() == 1);
+
+    {
+      Value result = callOp.getResult(0);
+      assert(result.use_empty() && "It is not allowed to return an iterator.");
```

Comment:
`It` ? maybe improve the error message ?

Reply:
OK, fixed in next PR.